subsys/testsuite/ztest/src/ztest_shell.c

```diff
@@ -0,0 +1,416 @@
+		.val = 'o',
+	},
+	{0},
+};
+static const char *const cmd_ztest_help =
+	"Run Ztest test suites\n"
+	"usage: ztest [arguments..]\n"
+	"\n"
+	"Without specifying any arguments, all tests from all testsuites are run.\n"
+	"\n"
+	"arguments:\n"
+	"--" PREFIX "filter=POSTIVE_PATTERNS[-NEGATIVE_PATTERNS]\n"
+	"                         run test that match POSITIVE_PATTERNS and do not match\n"
+	"                         NEGATIVE_PATTERNS\n"
+	"--help                   print this help message\n"
+	"--" PREFIX "list_tests   list all test suites and tests\n"
+	"--" PREFIX "also_run_disabled_tests\n"
+	"                         also run disabled tests\n"
+	"--" PREFIX "repeat       run tests repeatedly (use a negative number to repeat forever)\n"
+	"--" PREFIX "shuffle      randomize test order on every iteration\n"
+	"--" PREFIX "random_seed  seed for shuffling ([1,9999], 0 to seed based on current time)\n"
+	"--" PREFIX "color=(yes|no|auto)\n"
+	"                         enable / disable colored output. the default is off\n"
+	"--" PREFIX "print_time=(0|1)\n"
+	"                         print time duration for each test (defaults to 1)\n"
+	"--" PREFIX "output=(json|xml)[:DIRECTORY_PATH/|:FILE_PATH]\n"
+	"                         generate a json or xml report in the given directory or with\n"
+	"                         the given filename. The default is not to generate a report\n"
+	"\n"
+	"See https://pastebin.com/GtBg0aHZ for details\n";
+
+static int ztest_shell_parse_opts(const struct shell *sh, int argc, char *const *argv,
+				  struct ztest_shell_state *opts)
+{
+	int ret;
+	int option_index = 1;
+
+	while ((ret = getopt_long(argc, argv, "", ztest_shell_longopts, &option_index)) != -1) {
+		switch (ret) {
+		case 'h':
+			opts->help = true;
+			break;
+		case 'l':
+			opts->list_tests = true;
+			break;
+		case 'f':
+			opts->filter = optarg;
+			break;
+		case 'd':
+			opts->disabled = true;
+			shell_warn(sh, "running disabled tests is not yet implemented");
+			break;
+		case 'r':
+			opts->repeat = atoi(optarg);
+			break;
+		case 'u':
+			opts->shuffle = true;
+			shell_warn(sh, "shuffle is not yet implemented");
+			break;
+		case 's':
+			opts->seed = atoi(optarg);
+			shell_warn(sh, "random_seed is not yet implemented");
+			break;
+		case 'c':
+			opts->color = strcmp(optarg, "no") != 0;
+			shell_warn(sh, "color is not yet implemented");
+			break;
+		case 't':
+			opts->time = (bool)atoi(optarg);
+			shell_warn(sh, "print_time is not yet implemented");
+			break;
+		case 'o':
+			opts->output = optarg;
+			shell_warn(sh, "writing output to file is not yet implemented");
+			break;
+		case '?':
+		default:
+			return -EINVAL;
+		}
+	}
+
+	return 0;
+}
+
+static int ztest_shell_verify_opts(const struct shell *sh, struct ztest_shell_state *opts)
+{
+	ARG_UNUSED(sh);
+	ARG_UNUSED(opts);
+
+	return 0;
+}
+
+static int cmd_ztest(const struct shell *sh, size_t argc, char *const *argv)
+{
+	int ret;
+
+	_state = (struct ztest_shell_state){
```

Comment:
aren't `_` prefix variables reserved?

Reply:
single underscore followed by a lowercase is fine - the intent is that the symbols should only be visible within local scope / compilation unit. I usually use them to shadow opaque pointers of the same name.